Bug 10452 - Page Load Time Test Runner

shinglyu opened the pull request:
This is a minimal test runner for bug #10452 . 

The main entry point is `test_all.sh`, which will trigger `runner.py` (run the performance test), then run `submit_to_perfherder.py` (push performance data to perfherder)

There are also some utility script scattered around.

You can also check `README.md` for detail.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11107)

<!-- Reviewable:end -->

shinglyu committed:
Bug 10452 - Page Load Time Test Runner

highfive commented:
Heads up! This PR modifies the following files:
 * @aneeshusa: etc/ci/performance/git_log_to_json.sh, etc/ci/performance/master.cfg, etc/ci/performance/credential.json.example, etc/ci/performance/test_all.sh, etc/ci/performance/test_submit_to_perfherder.py, etc/ci/performance/prepare_manifest.sh, etc/ci/performance/.gitignore, etc/ci/performance/test_runner.py, etc/ci/performance/submit_to_perfherder.py, etc/ci/performance/README.md, etc/ci/performance/user-agent-js/01.perf-timing.js, etc/ci/performance/runner.py, etc/ci/performance/user-agent-js/02.timeout.js

shinglyu commented:
cc @autrilla

autrilla commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,172 @@
+    print("Running test:")
+    print(test_cmd)
+    print("Timeout:{}".format(timeout))
+    log = ""
```
Why aren't you simply returning subprocess.check_output here?

autrilla commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,172 @@
+    return [tc for tc in result_json if tc['testcase'] in manifest]
+
+
+def median(lst):
```
https://docs.python.org/3/library/statistics.html#statistics.median does this already, the downside is afaik it's not available for python 2, but since this is python 3, it should be fine

autrilla commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,321 @@
+    suites = []
+    measurement = "domComplete"  # Change this to an array when we have more
+
+    def getTimeFromNavStart(timings, measurement):
```
uber-nit: snake-case the method name?

aneeshusa commented:
I skimmed this and I'd like to make some review comments in a few days when my final projects are turned in.

highfive commented:
New code was committed to pull request.

shinglyu commented:
@autrilla I'm updating the README and runner in this patch. I might now merge it back to the servo-perf repo right now, so check here for updates.

highfive commented:
New code was committed to pull request.

shinglyu commented:
@aneeshusa r?

larsbergstrom commented:
@aneeshusa Do you have feedback on this? Otherwise, I'm going to land it and we can follow up with any changes you'd like.

aneeshusa commented on etc/ci/performance/.gitignore:
```diff
@@ -0,0 +1,9 @@
+page_load_test/*
```
No need to include this file here.

aneeshusa commented on etc/ci/performance/git_log_to_json.sh:
```diff
@@ -0,0 +1,5 @@
+# git log -n 1 --pretty=format:'{%n  "commit": "%H",%n  "abbreviated_commit": "%h",%n  "tree": "%T",%n  "abbreviated_tree": "%t",%n  "parent": "%P",%n  "abbreviated_parent": "%p",%n  "refs": "%D",%n  "encoding": "%e",%n  "subject": "%s",%n  "sanitized_subject_line": "%f",%n  "body": "%b",%n  "commit_notes": "%N",%n  "verification_flag": "%G?",%n  "signer": "%GS",%n  "signer_key": "%GK",%n  "autho ...
+
+# Removed body to prevent multiline & unescaped body string
```
Both of these comments can be removed.

aneeshusa replied:
Well, we should keep the second one but reword it a little, maybe
```
Don't include body to prevent multiline & unescaped body string
```

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,321 @@
+import argparse
```
File needs a shebang because it is a script.

aneeshusa commented:
Sorry about the long delay. Can you update this to include your new changes from the servo-perf repository?

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,171 @@
+    except subprocess.CalledProcessError as e:
+        print("Unexpected Fail:")
+        print(e)
+        print("You man want to re-run the test manually:\n{}".format(test_cmd))
```
spelling: may

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,171 @@
+    if len(lst) < 1:
+        return None
+    if len(lst) % 2 == 1:
+        return lst[int(((len(lst)+1)/2)-1)]
```
This can be `lst[len(lst)//2]`.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,171 @@
+    if len(lst) % 2 == 1:
+        return lst[int(((len(lst)+1)/2)-1)]
+    else:
+        return float(sum(lst[int((len(lst)/2)-1):int((len(lst)/2)+1)]))/2.0
```
Let's split this up into two lines:
```
middle = len(lst)//2
return sum(lst[upper - 1:upper + 1])/2

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,171 @@
+
+def save_result_json(results, filename, manifest, expected_runs):
+
+    print(results)
```
I'm assuming these are debugging prints; if so, we can remove them.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash -e
+export DISPLAY=:0 # In servo-linux1
```
Prefer to set this as an environment variable (i.e. from Buildbot).

aneeshusa replied:
Also, please update the shebang and add robustness options at the top of the script as outlined in the [style guide](https://github.com/servo/servo/blob/master/docs/STYLE_GUIDE.md).

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,321 @@
+
+        suite["subtests"].append({
+            "name": format_testcase_name(testcase["testcase"]),
+            "value": value}
```
style nit: put closing brace on the next line with the `)`

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,321 @@
+    suites.append(suite)
+
+    return (
+        {
```
style nit: opening brace can go on the previous line right after the `(`

aneeshusa replied:
Actually, we can get rid of the outer set of parentheses here entirely.

aneeshusa commented:
It looks like some of my review comments have already been taken care of in the servo-perf repo (like replacing median with the stdlib version), so I'll hold off on more review until you update this.

shinglyu commented:
@aneeshusa Sorry for taking so long. I've synced up with the `servo-perf` repo. I'm not planning to add any new feature right now, so we can try to merge this and deprecate the `servo-perf` repo. After that the contributors can work on the features in the servo/servo repo directly.

aneeshusa commented:
@shinglyu Thanks for the update, I have a few high-level questions. Do you want this to run on our infrastructure? If so, can you provide an estimate of how long the tests take to run (for scheduling purposes)? I see you're comparing against Firefox Nightly - would you like to see us comparing fresh Servo and Firefox nightlies each night (automated)? (It looks like the Treeherder graph is already recording revisions for each test, which would have been my next question.)

cc @larsbergstrom:  does moving to Gecko CI affect if we run these on our infra or Firefox infra?

larsbergstrom commented:
The initial thoughts I've seen on moving to shared / Gecko CI are that it *should* only affect the homu instance. That is, we'd need to have a version of homu that handles just servo/servo that lives within Mozilla platform engineering.

It's unclear whether any of our other services would move in, but even in the most extreme case, I'd only expect to see our CI builds and not our nightly or test builders (as this issue discusses) moving into their hosting for now.

Such a move would be much more compelling once TaskCluster supports more of our scenarios and we could get off of buildbot entirely :-)

shinglyu commented:
@aneeshusa 
* __Run in Servo infra?__ : I hope so. Also it will smooth the workflow in my own server right now. Currently I do the following:
```
git pull servo => build => copy the servo binary to a separate servo-perf checkout => run test
```
But if I need to update servo-perf, I'll have to do it manually. If we put it in tree, we can

```
git pull servo => build => run the in-tree servo-perf test
```

* __Time it takes__: ~ 1 hr for Servo and ~30 min for Gecko. But most of the time is wasted waiting for the test to time out (due to #11087 ). If we fix #11087 and our performance are comparable to Gecko, it will take ~30 min sequentially. There is no dependency between tests so we can split them into bins and run in parallel, so probably ~5 min for 6 bins? But we have to ensure each test environment has similar (virtual) hardware setup since we are testing performance.

* __Servo v.s. Gecko__: Yes, the ideal case is to compare Servo Nightly with Firefox Nightly. But I don't have a robust way to update Firefox Nightly and get the revision data (needed for Perfherder graphing). And also Servo has a fixed SpiderMonkey version, so we fixed the Firefox version to 48.0a2 for now. I believe eventually we'll update them daily.

aneeshusa commented:
OK, for now I'll focus on reviewing the actual test runner and we can keep running this on your server; afterwards we can see about moving to our infrastructure. It sounds like running this nightly won't be too bad, even if we don't parallelize. AFAIK all of our `servo-linux*` machines are the same EC2 instance type and configured in the same way, so that should be taken care of.

aneeshusa commented on etc/ci/performance/user-agent-js/01.perf-timing.js:
```diff
@@ -0,0 +1,33 @@
+function formatLine(name, t){
+  var output = "[PERF]," + name + "," + t;
+  console.log(output);
+  //document.getElementById('timing').innerHTML += output + "<br/>";
```
Looks like this comment can be removed, and `output` can be inlined.

aneeshusa commented:
Can you make another python module in `etc/ci/` that invokes all of the unit tests (I believe that's `test_runner.py` and `test_submit_to_perfherder.py`? This will make it easier to run the tests as part of the regular Servo build + test suite.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,68 @@
+Servo Page Load Time Test
+==============
+
+[Tracking Bug](https://github.com/servo/servo/issues/10452)
```
This won't be needed any more.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,68 @@
+
+[Tracking Bug](https://github.com/servo/servo/issues/10452)
+
+# Prerequisite
```
nit: Prerequiste -> Prerequisites (in case there are ever more)

aneeshusa commented:
What is the purpose of the `firefox/profile.tar.gz` file? I extracted it locally and it seems to mostly be cache items and history visits, neither of which I understand the need to include.

Cache items: we're serving the pages locally to remove network variability, and it doesn't seem fair to pre-fill the Firefox cache but not the Servo cache. I would imagine we would start from a fresh (empty) cache for each test; otherwise, we could just load the page a few times first to warm the cache and ignore those runs when calculating the result, without including those items in the repo. (Also, not sure if there are potential copyright concerns with distributing those files; I didn't look at the cache contents.) Keeping cache items in the repo doesn't seem useful to me.

Similarly, I don't see why we should include a history of pages visited.

In general including binary files in the repo is a code smell for me, so if it is important to have these files, maybe we can put them on S3 or similar (not sure about versioning/how much they would chnage).

aneeshusa commented on etc/ci/performance/test_submit_to_perfherder.py:
```diff
@@ -0,0 +1,187 @@
+            ]
+        }
+    }
+    '''
```
This entire string seems to be unused and can be removed.

aneeshusa commented on etc/ci/performance/test_submit_to_perfherder.py:
```diff
@@ -0,0 +1,187 @@
+
+    expected = {
+        "performance_data": {
+            # TODO: can we create a framwork on treeherder
```
spelling: framwork -> framework

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,68 @@
+# Running in buildbot
+
+* We have a `master.cfg` for you to run it in buildbot. You can use the `servo-linux1` vagrant VM setup from [servo/saltfs](https://github.com/servo/saltfs) to run it.
+* You'll need to setup this repository manully in the VM, check the inline comments in `master.cfg` for detail.
```
spelling: manully -> manually

aneeshusa replied:
Actually, it looks like the entire `Running in buildbot` section can be removed since we're not including the `master.cfg`.

aneeshusa commented on etc/ci/performance/.gitignore:
```diff
@@ -0,0 +1,8 @@
+
+.ropeproject
+__pycache__/
+*.pyc
```
This is in the top-level `.gitignore` already.

aneeshusa commented on etc/ci/performance/credential.json.example:
```diff
@@ -0,0 +1,6 @@
+  "client_id":"your_id",
+  "secret": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
+  "S3_ACCESS_KEY": "xxxxxxxxxxxxx",
+  "S3_SECRET_KEY": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```
The S3 secrets don't seem to be used anywhere.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,381 @@
+    client = TreeherderClient(protocol='https',
+                              host='treeherder.allizom.org',
+                              # protocol='http',
+                              # host='local.treeherder.mozilla.org',
```
These two lines can be removed.

aneeshusa commented on etc/ci/performance/credential.json.example:
```diff
@@ -0,0 +1,6 @@
+{
+  "client_id":"your_id",
+  "secret": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
```
I would prefer to pass these as environment variables, e.g. `TREEHERDER_CLIENT_ID` and `TREEHERDER_CLIENT_SECRET`.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,68 @@
+
+# Upload log to S3
+* `pip install boto3`
+* `python3 upload_log_to_s3.py <filename>`
```
This section can also be removed since `upload_log_to_s3.py` is not included either.

aneeshusa commented:
Is there any way to unify `firefox/addon/data/perf.js` with `user-agent-js/01.perf-timing.js` and `user-agent-js/02.timeout.js`?

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+
+def get_servo_command(url, timeout):
+    ua_script_path = "{}/user-agent-js".format(os.getcwd())
+    test_cmd = ("timeout {timeout}s ./servo/servo '{url}'"
```
Since this is in-tree, the easiest way to get a servo binary is `./mach build` (`--release` I assume). After that with the PWD being the top level of the servo repo, we can just use `target/release/servo` instead of requiring copying the `servo` binary to `etc/ci/performance/servo/servo`.

shinglyu replied:
I'm planning to add a `./mach test-perf` command, can I change that path in that followup patch?

aneeshusa replied:
Sure, changing the path later is fine.

Ditto about returning a list here instead of relying on implicit word splitting.

aneeshusa commented:
The only fields I am seeing getting used from the `revision.json` files are:
- `author.name`
- `author.email`
- `author.timestamp`
- `commit`
- `subject`

If we don't need any of the other fields, let's remove them.

shinglyu commented:
Review status: 0 of 19 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[etc/ci/performance/runner.py, line 89 \[r4\]](https://reviewable.io:443/reviews/servo/servo/11107#-KHP5GGvWIKVcG-2kpVh-r4-89:-KMMylpCaSS5Z6czTvvz:-1071938725) ([raw file](https://github.com/servo/servo/blob/ff3928a3bd9e269024efb52300c1b356594e954a/etc/ci/performance/runner.py#L89)):*
<details><summary><i>Previously, aneeshusa (Aneesh Agrawal) wrote…</i></summary>
> This can be `lst[len(lst)//2]`.
</details>
replaced by python3 built-in median

---

*[etc/ci/performance/runner.py, line 91 \[r4\]](https://reviewable.io:443/reviews/servo/servo/11107#-KHP5GGvWIKVcG-2kpVh-r4-91:-KMMyqoEAj_Vkj5ctdiB:-173126001) ([raw file](https://github.com/servo/servo/blob/ff3928a3bd9e269024efb52300c1b356594e954a/etc/ci/performance/runner.py#L91)):*
<details><summary><i>Previously, aneeshusa (Aneesh Agrawal) wrote…</i></summary>
> Let's split this up into two lines:
> 
> ```
> 
> middle = len(lst)//2
> 
> return sum(lst[middle - 1:middle + 1])/2
</details>
replaced by python3 built-in median

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11107)*
<!-- Sent from Reviewable.io -->

shinglyu commented:
The `profile.tar.gz` is for running Firefox with the add-on pre-installed. It's kind of a hack, and it's not as automated as I believe (still need some x-forwarding into the VM and click on a few buttons), so I'll remove it and add the instructions to the README.

I also didn't automate the Firefox download, setup and upgrade part, so setting up Firefox manually is probably the way to go right now.

shinglyu commented:
My tests are written using the `py.test` framework. It has automatic test discovery, so how would you suggest me to write the test module?

shinglyu commented:
Sorry for the delay, I've addressed most of the review comments. Let me know your opinion on the rest of the issues :)

aneeshusa commented on etc/ci/performance/.gitignore:
```diff
@@ -0,0 +1,7 @@
+servo/*
+output.png
+output/*
+credential.json
```
No longer needed.

aneeshusa commented on etc/ci/performance/.gitignore:
```diff
@@ -0,0 +1,7 @@
+output/*
+credential.json
+
+.ropeproject
```
This appears to be for some refactoring tool? I think this can be removed.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,58 @@
+* Outside of vm, open `http://local.treeherder.mozilla.org` and login to create an account
+* `vagrant ssh`
+  * `./manage.py create_credentials slyu slyu@mozilla.com "description"`, the email has to match your logged in user. Remember to log-in through the Web UI once before you run this.
+  * Open a file called `credential.json`. Copy the clinet secrent to your `credential.json`. You can use `credential.json.example` as a template.
```
These instructions should be updated for environment variable usage.

spelling: clinet secrent -> client secret

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,58 @@
+## For Gecko
+
+* Install Firefox Nightly in your PATH
+* Install the `servo-perf-0.0.1.xpi` add-on to your Firefox Nightly
```
Can we generate the xpi from the files in the repo instead of including the actual xpi?

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,58 @@
+
+# Unit tests
+
+The following tests can be run by `python -m pytest <filename>`:
```
Would be nice to mention the `py.test` auto-discovery here to run these all in one command.

aneeshusa commented on etc/ci/performance/firefox/revision_46.0.1.json:
```diff
@@ -0,0 +1,28 @@
+{
```
The `package.json` for the Firefox addon says version `>=38.0a1` but this filename is `revision_46.0.1.json`. Which version is needed? We should also mention the version to use in the README (instead of just `Nightly`).

Also, let's pare down the fields to just what's needed, the same as what `git_log_to_json.sh` outputs.

shinglyu replied:
`package.json` is the minimal required version. Here we put a the 46.0.1 json because we are currently comparing Servo with Gecko 46.0.1. I guess we can remove this because I don't want to support Gecko testing in servo/servo right now.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+
+
+def get_gecko_command(url, timeout):
+    test_cmd = ("timeout {timeout}s ./firefox/firefox/firefox"
```
Let's return a list instead of relying on implicit word splitting.

aneeshusa commented on etc/ci/performance/user-agent-js/01.perf-timing.js:
```diff
@@ -0,0 +1,51 @@
+    if (window.dump) {
+      window.dump(o + '\n');
+    }
+    //window.dump(o + '\n') // For Gecko
```
Can be removed.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+        tj.add_option_collection(data['job']['option_collection'])
+
+        # for log_reference in data['job']['log_references']:
+        #    tj.add_log_reference( 'buildbot_text', log_reference['url'])
```
Can be removed.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+        elif copy:
+            block.append(line)
+
+    placeholder = {
```
I think there used to be a comment explaining all the `-1`s, can you please put it back?

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+        for line in block:
+            try:
+                key = line.split(",")[1]
+                value = line.split(",")[2]
```
Use destructuring to extract `key` and `value` at the same time.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+                    {
+                        'type': 'json',
+                        'name': 'performance_data',
+                        # 'job_guid': job_guid,
```
Can this be removed or is this a TODO?

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+                        #            ]
+                        #        }]
+                        #     }
+                        # }
```
Can be removed.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+    parser = argparse.ArgumentParser(
+        description=("Submit Servo performance data to Perfherder. "
+                     "Put your treeherder credentail in credentail.json. "
+                     "You can refer to credential.json.example.")
```
Please update to mention using environment variables instead.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+    perf_data = format_perf_data(result_json, args.engine)
+    failures = list(filter(lambda x: x['domComplete'] == -1, result_json))
+    summary = format_result_summary(result_json)
+    summary = summary.replace('\n', '<br/>')
```
I think this might fit on the previous line.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+    trsc = create_resultset_collection(dataset)
+
+    result = "success"
+    # FIXME: Always passing until https://bugzil.la/1276178 is fixed
```
Looks like this bug has been fixed for a month (in master at least, not sure if it's been pushed out).

shinglyu replied:
I remember Perfherder has other bad behavior on failed tests. I'll update the comment to reflect that.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+        # for log_reference in data['job']['log_references']:
+        #    tj.add_log_reference( 'buildbot_text', log_reference['url'])
+
+        # data['artifact'] is a list of artifacts
```
Please remove this comment.

aneeshusa commented on etc/ci/performance/firefox/revision_46.0.1.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "commit": "0b8492c110be",
```
Should this be a full (length 40) hash? Possibly related to https://bugzilla.mozilla.org/show_bug.cgi?id=1283856.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,380 @@
+
+def format_testcase_name(name):
+    temp = name.replace('http://localhost:8000/page_load_test/', '')
+    temp = temp.replace('http://localhost:8000/tp6/', '')
```
This looks suspicious, should this be `tp5` instead of `tp6`?

shinglyu replied:
tp6 is my experimental test suite, which I haven't turn on by default. TP5 is from 2011, I think we can build a TP6 for 2016, but not ready yet.

aneeshusa replied:
Please add a comment to that effect so it's clear that it's not a typo.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,58 @@
+* Python3
+
+# Usage
+## Build Servo
```
Please put this section after `Prepare the test runner`, as that only needs to be done once but this has to be re-done for each Servo commit (nightly).

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+                print("Running test {}/{} on {}".format(run + 1,
+                                                        args.runs,
+                                                        testcase))
+                command = command_factory(testcase, args.timeout)
```
Hoist this above the `for run in ...` loop to avoid calling `command_factory` many times.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+                command = command_factory(testcase, args.timeout)
+                log = execute_test(testcase, command, args.timeout)
+                result = parse_log(log, testcase)
+                # TODO: check for other measurements
```
Please provide more details (which other measurements)?

shinglyu commented:
@aneeshusa ping

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,66 @@
+* Outside of vm, open `http://local.treeherder.mozilla.org` and login to create an account
+* `vagrant ssh`
+  * `./manage.py create_credentials slyu slyu@mozilla.com "description"`, the email has to match your logged in user. Remember to log-in through the Web UI once before you run this.
+  * Open a file called `credential.json`. Copy the client secret to your `credential.json`. You can use `credential.json.example` as a template.
```
We're no longer using `credential.json`.

aneeshusa commented on etc/ci/performance/prepare_manifest.sh:
```diff
@@ -0,0 +1,8 @@
+set -o pipefail
+
+# TP5 manifest uses `localhost`, but our local server probably don't use port 80
+sed 's/localhost\/page_load_test\/tp5n/localhost:8000\/page_load_test/g' ./page_load_test/tp5o.manifest > ./page_load_test/tp5o_8000.manifest
```
Please split this onto multiple lines for readability.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,263 @@
+
+def get_servo_command(url, timeout):
+    ua_script_path = "{}/user-agent-js".format(os.getcwd())
+    test_cmd = ["timeout", "{timeout}s".format(timeout),
```
We can just return this directly.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,263 @@
+
+
+def get_gecko_command(url, timeout):
+    test_cmd = ["timeout", "{timeout}s".format(timeout),
```
This can also be returned directly.

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,263 @@
+        testcases = load_manifest(args.tp5_manifest)
+        results = []
+        for testcase in testcases:
+            command = command_factory(testcase, args.timeout)
```
Since we're running both commands under `timeout` in the same way, let's handle wrapping the command with `timeout` here instead of inside `get_servo_command` and `get_gecko_command`.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,363 @@
+    trsc = TreeherderResultSetCollection()
+
+    for data in dataset:
+
```
Remove this empty line.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,363 @@
+
+        revisions = []
+        for rev in data['revisions']:
+
```
Remove this empty line.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,363 @@
+            tr.add_repository(rev['repository'])
+
+            revisions.append(tr)
+
```
Remove this empty line.

aneeshusa commented on etc/ci/performance/submit_to_perfherder.py:
```diff
@@ -0,0 +1,363 @@
+    tjc = TreeherderJobCollection()
+
+    for data in dataset:
+
```
Remove this empty line.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,22 @@
+
+source venv/bin/activate
+
+echo "Staring the local server"
```
Staring -> Starting

aneeshusa commented on etc/ci/performance/test_all_gecko.sh:
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
```
Let's combine this script with `test_all.sh` and have it take a parameter for the engine to use.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,66 @@
+## Prepare the test runner
+
+* Clone this repo
+* Download [tp5n.zip](http://people.mozilla.org/~jmaher/taloszips/zips/tp5n.zip), extract it to `page_load_test/`
```
If you're planning to create a `tp6

aneeshusa replied:
Actually, after looking at `prepare_manifest.sh`, we can use `page_load_test/tp5n/`, which will cut down the noise in the file and still allow for multiple test suites.

aneeshusa commented on etc/ci/performance/user-agent-js/01.perf-timing.js:
```diff
@@ -0,0 +1,50 @@
+    }
+}
+
+function formatLine(name, t){
```
nit: space before `{`

aneeshusa commented on etc/ci/performance/user-agent-js/01.perf-timing.js:
```diff
@@ -0,0 +1,50 @@
+  print("[PERF]," + name + "," + t);
+}
+
+function printPerfTiming(){
```
nit: space before `{`

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+    engine="--engine gecko"
+    ;;
+  *)
+    echo "Can't understand the argument, using the default servo engine"
```
Instead of defaulting to `servo` in this case, we should just error out.
It would be nice to handle `--help` as well and print usage, but that's not necessary.

shinglyu replied:
I'll probably wrap that in `mach`, so probably don't need `--help`.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+set -o nounset
+set -o pipefail
+
+case "$1" in
```
`set -o nounset` will make this an error in case no arguments are passed, but the usage indicates this should be allowed (IIUC defaulting to the `servo` engine).

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+
+case "$1" in
+  --servo)
+    engine=""
```
Might as well do "--engine servo" for explicit-ness.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+
+echo "Running tests"
+python3 runner.py --runs 3 $MANIFEST $PERF_FILE 
+python3 runner.py $engine --runs 3 $MANIFEST $PERF_FILE
```
Running `runner.py` twice with different arguments seems like a bug.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+
+#MANIFEST="page_load_test/test.manifest"
+MANIFEST="page_load_test/20160509.manifest" # A manifest that excludes timeout test cases
+PERF_FILE="output/perf-$(date +"%s").json"
```
It looks like this filename can be changed to whatever we like, so I'd prefer to use "date --iso-8601=seconds" as being more human-friendly than epoch time. If not, IMO we can drop the quotes around `%s`.

aneeshusa commented:
I think a rebase (you can squash as well) will take care of the tidy failure. Also, take a look at the linting being added in #12295 and please make sure the scripts conform to it as I expect that to be merged soon.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+    ;;
+esac
+
+source venv/bin/activate
```
When we run this on the builders we'll likely either already be inside the virtualenv from `mach` or have system-wide-installed libraries available, so please remove this line. Probably would be good to update the docs to mention this.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+echo "Starting the local server"
+python3 -m http.server > /dev/null 2>&1 &
+
+#MANIFEST="page_load_test/test.manifest"
```
I'm guessing this is meant to be uncommented when #11087 is fixed - can you add a TODO comment with a link?

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+PERF_FILE="output/perf-$(date +"%s").json"
+
+echo "Running tests"
+python3 runner.py --runs 3 $MANIFEST $PERF_FILE 
```
Please quote all variables and use the full "${variable}" form throughout the script, per the [style guide](https://github.com/servo/servo/blob/master/docs/STYLE_GUIDE.md).

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+echo "Running tests"
+python3 runner.py --runs 3 $MANIFEST $PERF_FILE 
+python3 runner.py $engine --runs 3 $MANIFEST $PERF_FILE
+sudo ntpdate tw.pool.ntp.org
```
Why do we need this? Not all systems have or use `ntpdate` and I'd prefer to remove this line and instead document the fact that NTP should be enabled.

aneeshusa commented on etc/ci/performance/test_all.sh:
```diff
@@ -0,0 +1,38 @@
+python3 runner.py $engine --runs 3 $MANIFEST $PERF_FILE
+sudo ntpdate tw.pool.ntp.org
+echo "Submitting to Perfherder"
+# XXX: we are using servo's revision to make the graph pretty
```
What is the `XXX` for?

aneeshusa commented on etc/ci/performance/runner.py:
```diff
@@ -0,0 +1,260 @@
+        testcases = load_manifest(args.tp5_manifest)
+        results = []
+        for testcase in testcases:
+            command = ["timeout", "{timeout}s".format(args.timeout)] +
```
This either needs a `\` at the end to signal the line continuation, or a temporary variable.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,66 @@
+* Clone this repo
+* Download [tp5n.zip](http://people.mozilla.org/~jmaher/taloszips/zips/tp5n.zip), extract it to `page_load_test/`
+* Run `prepare_manifest.sh` to transform the tp5n manifest to our format
+* `python3 -m virtualenv venv; source venv/bin/activate; pip install treeherder-client`
```
Instead of requiring a virtualenv, let's do something along the lines of:
```
Install the Python3 treeherder-client library.
For example, to install in a virtualenv, do: [snip]
```

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,66 @@
+  * `./bin/run_gunicorn`
+* Outside of vm, open `http://local.treeherder.mozilla.org` and login to create an account
+* `vagrant ssh`
+  * `./manage.py create_credentials slyu slyu@mozilla.com "description"`, the email has to match your logged in user. Remember to log-in through the Web UI once before you run this.
```
I'm guessing you don't want to hard-code `slyu` and `slyu@mozilla.com` here.

aneeshusa commented on etc/ci/performance/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## Test Perfherder Locally
+
+If you want to test the data submission code in `submit_to_perfherder.py`, you can setup a local treeherder VM. If you don't need to test `submit_to_perfherder.py`, you can skip this step.
```
Is this needed to run `test_submit_to_perfherder.py`? Please add a note either way.

metajack commented:
Why is python3 required for this? Can't we use python2.7 with `from future import`s? I really don't want to require multiple versions of python for our build.

shinglyu commented:
@metajack I don't expect to run this in the build process in the near future. Checking this into the tree is for people to run on their own computer, so I believe this won't impact our CI very much. We actually migrated from 2.7 to 3 because of previous review comments, so may we land this first and discuss the CI integration later?

metajack commented:
@shinglyu That sounds fine.

larsbergstrom commented:
It looks to me like this has been fully reviewed and is ready to go except for license issues reported by tidy:
```
./etc/ci/performance/runner.py:1: incorrect license

./etc/ci/performance/submit_to_perfherder.py:1: incorrect license

./etc/ci/performance/test_runner.py:1: incorrect license

./etc/ci/performance/test_submit_to_perfherder.py:1: incorrect license

```

shinglyu commented:
Just rebased to the latest master and clean up all the tidy errors.

KiChjang commented:
@shinglyu Why did you manually merge this?

shinglyu commented:
Oh no! I subconsciously click the merge button. Is there any way to revert this?

shinglyu commented:
I'm so sorry, I hope I didn't break any Ci stuff.

metajack commented:
If master is still green, I think it's fine to let this stand. If we have broken the three we should revert this or quickly fix in a followup.

shinglyu commented:
This does not touch any servo code, its some self-contained python code in the `/etc/ci` directory, so I believe the tree is still green.

metajack commented:
Ok, no harm done then since this had already been reviewed.

We're looking into ways to prevent accidental merges like this. It seems GitHub may have added some new things since we last tried to do this.
